### PR TITLE
Chapter 12: Reset password

### PR DESCRIPTION
**Pull request by hazwanizafirah**

### CODE REVIEW

**Comment by hazwanizafirah** [comment]
image
When I click the link to reset password, it shows this error. Can I know how to fix this?

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -12,10 +12,9 @@ def create
         message = "Account not activated. Check your email for the activation link."
         flash[:warning] = message
         redirect_to root_url
-      end
```
Comment (huongnt-2545) [inline app/controllers/sessions_controller.rb:15]: keep the previous code

File: app/helpers/password_resets_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
```
Comment (huongnt-2545) [inline app/helpers/password_resets_helper.rb:1]: delete this file, if it is not useful

File: app/views/password_resets/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<% provide(:title, 'Reset password') %>
+<h1>Reset password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
+      <%= render 'shared/error_messages' %>
```
Comment (huongnt-2545) [inline app/views/password_resets/edit.html.erb:7]:
```suggestion
      <%= render 'shared/error_messages', object: f.object %>
```

File: test/controllers/account_activations_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
```
Comment (huongnt-2545) [inline test/controllers/account_activations_controller_test.rb:1]: delete this file

File: test/mailers/previews/user_mailer_preview.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
```
Comment (huongnt-2545) [inline test/mailers/previews/user_mailer_preview.rb:1]: delete this file

File: test/mailers/user_mailer_test.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+require "test_helper"
```
Comment (huongnt-2545) [inline test/mailers/user_mailer_test.rb:1]: delete this file

File: test/controllers/password_resets_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+require "test_helper"
```
Comment (huongnt-2545) [inline test/controllers/password_resets_controller_test.rb:1]: delete this file

**Comment by hazwanizafirah** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+  def new
+  end
```
Comment (anlvq-1058) [inline app/controllers/password_resets_controller.rb:6]:
```suggestion
  def new; end
```

File: app/controllers/password_resets_controller.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update]
+  before_action :valid_user, only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
```
Comment (anlvq-1058) [inline app/controllers/password_resets_controller.rb:4]:
```suggestion
  before_action :get_user, :valid_useronly, :check_expiration, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  def update
+    if params[:user][:password].empty? # Case (3)
+     @user.errors.add(:password, "can't be empty")
+     render 'edit'
+    elsif @user.update(user_params) # Case (4)
+      log_in @user
+      flash[:success] = "Password has been reset."
+      redirect_to @user
+    else
+      render 'edit' # Case (2)
```
Comment (anlvq-1058) [inline app/controllers/password_resets_controller.rb:33]:
```suggestion
      render :edit # Case (2)
```

File: app/controllers/password_resets_controller.rb [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+      flash[:info] = "Email sent with password reset instructions"
+      redirect_to @user
+    else
+      flash.now[:danger] = "Email address not found"
+      render :new
+    end
+  end
+
+  def edit
+  end
```
Comment (anlvq-1058) [inline app/controllers/password_resets_controller.rb:22]:
```suggestion
  def edit; end
```

File: app/controllers/password_resets_controller.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    return if (@user && @user.activated && @user.authenticated?(:reset, params[:id]))
+    redirect_to root_url
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
+    flash[:danger] = "Password reset has expired."
+    redirect_to new_password_reset_url
+  end
+end
```
Comment (anlvq-1058) [inline app/controllers/password_resets_controller.rb:57]: indent please

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
ready

File: config/routes.rb [hunk 13]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+  get 'password_resets/new'
+  get 'password_resets/edit'
+  resources :password_resets, only: [:new, :create, :edit, :update]
```
Comment (anlvq-1058) [inline config/routes.rb:4]: push this line into scope

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
Hi, I accidentally force update and git add . git commit --amend. git push origin chapter_12 -f. then all my coding in chapter 13 go to chapter 12 branch. is there any way for me to undo after force the update?

**Comment by hazwanizafirah** [comment]
ready

File: app/assets/stylesheets/microposts.scss [hunk 14]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
+// They will automatically be included in application.css.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/microposts.scss:2]: delete this file

File: app/controllers/password_resets_controller.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      flash.now[:danger] = t(:email_not_found)
+      render :new
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty? # Case (3)
+     @user.errors.add(:password, "can't be empty")
```
Comment (ngocvt-0484) [inline app/controllers/password_resets_controller.rb:23]: wrong indent
2spaces please

**Pull request closed without merging**